torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -142,29 +148,94 @@ def forward(
         H, D = self.heads, self.out_channels // self.heads
 
         k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
-
-        # Iterate over node-types:
-        for node_type, x in x_dict.items():
-            k_dict[node_type] = self.k_lin[node_type](x).view(-1, H, D)
-            q_dict[node_type] = self.q_lin[node_type](x).view(-1, H, D)
-            v_dict[node_type] = self.v_lin[node_type](x).view(-1, H, D)
-            out_dict[node_type] = []
-
-        # Iterate over edge-types:
-        for edge_type, edge_index in edge_index_dict.items():
-            src_type, _, dst_type = edge_type
-            edge_type = '__'.join(edge_type)
-
-            a_rel = self.a_rel[edge_type]
-            k = (k_dict[src_type].transpose(0, 1) @ a_rel).transpose(1, 0)
-
-            m_rel = self.m_rel[edge_type]
-            v = (v_dict[src_type].transpose(0, 1) @ m_rel).transpose(1, 0)
-
-            # propagate_type: (k: Tensor, q: Tensor, v: Tensor, rel: Tensor)
-            out = self.propagate(edge_index, k=k, q=q_dict[dst_type], v=v,
-                                 rel=self.p_rel[edge_type], size=None)
-            out_dict[dst_type].append(out)
+        if WITH_PYG_LIB:
+            # parralelize over node-types
+            node_types, xs = x_dict.keys(), x_dict.values()
+
+            # compute K
+            k_wts = [self.k_lin[node_type].weight for node_type in node_types]
+            k_biases = [self.k_lin[node_type].bias for node_type in node_types]
+            k_list = pyg_lib.ops.grouped_matmul(inputs=xs, others=k_wts,
```

Comment:
I would prefer to use `segment_matmul` here and first concatenate `x_dict` together. Otherwise, we are restricting ourselves to PyTorch 1.14/2.0 here which is not even released.